Add a Logger class to centralize Log Management

kamalca opened the pull request:
All print statements now use a centralized class to report errors, warnings, and other information.

<img width="584" alt="screen shot 2018-09-01 at 4 23 38 pm" src="https://user-images.githubusercontent.com/24720843/44950689-75d79e00-ae03-11e8-8eda-00b33c556eca.png">

Note: Some messaged from pods or from UI elements still do not use the error class because I did not want to edit any of the locked files that are in charge of printing these messages.

I did my best to clean up the error reporting so that any error could be traced back to the code that produced them.

As for reporting what view controller they came from, the error itself should be good enough to track which code the error came from. However, something that I I looked into and may be useful is reporting which ViewController is open when the error was reported. This feature could be added upon request.

kamalca committed:
Wrote ErrorLogger class

This error logger will be used to report all errors, warnings, info, etc. This is useful becasue it standardizes how information is reported to the output console and will allow error messages to be labled and ranked on importance.

kamalca committed:
Changed print statements to use the new Logging class

All print statements in the project have been made to be more useful and more organized by using the Logger class.

HirdayGupta commented on Bruin Bite/Models/ErrorLogger.swift:
```diff
@@ -0,0 +1,37 @@
+}
+
+class Logger {
+    static let shared = Logger()
```
@kamalca @ayush-patel Do you foresee the Logger class ever having a state of its own? (I.e. its own private data). It doesn't right now, and if that isn't something that we foresee, then we can change this from a Singleton class to just a class with static functions. (UPDATE: I don't think its ever going to have a state of its own in the foreseeable future, so go with the static approach please)

That change would involve changing all the (one) functions to static and removing the `static let shared` singleton (because all function calls would be through a static function).

This means that all calls to this class will now look like this: `Logger.handle(type: message:)`

HirdayGupta commented on Bruin Bite/Models/ErrorLogger.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    private init() {    }
+    
+    func handle(type:ErrorType, message:String) {
```
Suggest changing the method name & using function argument labels to modify the function signature to the following: (and using `static` to abide by the previous suggestion)
```swift
static func log(message: String, ofType type: ErrorType)
```
This and the switch from singleton to static would mean that calls to this function would now look like:
```swift
Logger.log(message: "Hi! I'm an error!", ofType: .error)
```
----
An alternate is dropping the label for message entirely, so the function definition would look like this:
```swift
static func log(_ message: String, withType type: ErrorType)
```
Now, this change and the switch from singleton to static would mean that calls to this function would now look like:
```swift
Logger.log("Hi! I'm an error!", withType: .error)
```

**I'm in favor of the second one** because the function name `log`, implies the verb, and hence, the label `message` for the first argument is kind of implied, because what else would you be logging lol. (EDIT: @kamalca let me know which alternative you prefer, and if you agree with the second one, go ahead and implement it!)

Side note: This may seem trivial, but think about it this way - the interface you are designing now is going to be used across the app for many more versions to come. Plus, our code could do with some beauty 😄

HirdayGupta commented on Bruin Bite/Models/ErrorLogger.swift:
```diff
@@ -0,0 +1,37 @@
+//
+//  ErrorLogger.swift
```
Rename the file to `LogManager.swift` and move it from the Models folder to a **new** folder called `Managers`.

HirdayGupta commented on Bruin Bite/Models/ErrorLogger.swift:
```diff
@@ -0,0 +1,37 @@
+
+import Foundation
+
+enum ErrorType: String {
```
Rename this to LogLevel (because names like ErrorType should be reserved for our Centralised Error Handling mechanism, whenever we get to it). 

When you do make this change, you should also replace the word 'type' from the argument name and labels in the `handle` function to 'level'.

HirdayGupta commented on Bruin Bite/Models/ErrorLogger.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    func handle(type:ErrorType, message:String) {
+        
+        if(message == "")
```
change to `message.isEmpty`

HirdayGupta commented on Bruin Bite/Models/ErrorLogger.swift:
```diff
@@ -0,0 +1,37 @@
+    case debug = "💙 -DEBUG- 💙"
+    case verbose = "💜 -VERBOSE- 💜"
+    case other = "🖤 -OTHER- 🖤"
+    case none = ""
```
While it's tempting to use `rawValue`s to determine the header for the different types here, that's not the intended use case for `rawValue`. I would instead add a function in the enum that returns the desired header string based on its own type and leave the raw values blank for now.

so the function would be added inside the enum and would look something like this:
```swift
func header() -> String {
    switch self {
        case .error:
            return "❤️ -ERROR- ❤️"
        // Handle all the other cases here (except maybe none, because that will be the same as default
        default:
            return ""
    }
}
```

HirdayGupta commented on Bruin Bite/Models/ErrorLogger.swift:
```diff
@@ -0,0 +1,37 @@
+            return
+        }
+        
+        let header:String = type.rawValue
```
With that `header()` function added into the enum, change this to:
```swift
let header: String = type.header()
```

HirdayGupta commented on Bruin Bite/Models/ErrorLogger.swift:
```diff
@@ -0,0 +1,37 @@
+import Foundation
+
+enum ErrorType: String {
+    case error = "❤️ -ERROR- ❤️"
```
Add documentation comments above each of these cases, such that XCode's autocomplete can help callers of the function will know when to use what log level. (the comments being added in will be displayed along with all the different available enum options in the autocomplete menu). 

The added comments + the change of "leaving the raw values blank" would mean the cases look like this:
```swift
/** Message describing an error that causes catastrophic failure. **/
case error
/** Message describing an error that causes a minor inconvenience. **/
case warning
/** Message describing some general information. **/
case info
/** Messages used solely for debug purposes. All DEBUG logs will be removed before pushing to production. **/
case debug
/** Lengthy messages used for detailed descriptions. **/
case verbose
/** Messages that don't fall under any of these categories (be very sure of this) **/
case other
/** Messages that don't want headers before their log outputs. **/
case none
```

HirdayGupta commented on Bruin Bite/Network/ChatAPI.swift:
```diff
@@ -32,12 +32,12 @@ class ChatAPI {
             }
         } else {
-            print ("No chat room label given")
+            Logger.shared.handle(type: .warning, message: "No chat room label given")
```
Change type to `.error` in light of recently added descriptions of the different log types. Justification: An empty chat room label would cause the chat room to fail and indicates something fundamentally wrong with our segue structure.

HirdayGupta commented on Bruin Bite/Network/MenuAPI.swift:
```diff
@@ -133,7 +133,12 @@ extension API{
                                         
                                     }
                                 }
-                                parsedLocation[Location(rawValue: location)!] = parsedItems
```
Great job identifying the force-unwrap here!

HirdayGupta commented on Bruin Bite/Network/MenuAPI.swift:
```diff
@@ -143,10 +148,10 @@ extension API{
                 }
             case let .failure(error):
-                print(error)
+                Logger.shared.handle(type: .error, message: "getOverviewMenu: \(error)")
```
Change message to "getOverviewMenu failed with error: \(error)"

HirdayGupta commented on Bruin Bite/Network/MenuAPI.swift:
```diff
@@ -278,10 +283,10 @@ extension API{
                 }
             case let .failure(error):
-                print(error)
+                Logger.shared.handle(type: .error, message: "getDetailedMenu: \(error)")
```
Same as previous. Change error message to "getDetailedMenu failed with error: \(error)"

HirdayGupta commented on Bruin Bite/Network/MenuAPI.swift:
```diff
@@ -317,10 +322,10 @@ extension API{
                 }
             case let .failure(error):
-                print(error)
+                Logger.shared.handle(type: .error, message: "Activity Levels: \(error)")
```
same as above.

HirdayGupta commented on Bruin Bite/Network/MenuAPI.swift:
```diff
@@ -352,10 +357,10 @@ extension API{
                 }
             case let .failure(error):
-                print(error)
+                Logger.shared.handle(type: .error, message: "getHours: \(error)")
```
same as above.

HirdayGupta commented on Bruin Bite/Network/UserAPI.swift:
```diff
@@ -29,13 +29,13 @@ extension API {
+                        Logger.shared.handle(type: .error, message: "createUser: \(err)")
                     } catch let reserr {
-                        print(reserr)
+                        Logger.shared.handle(type: .error, message: "createUser: \(reserr)")
```
Same for all these messages. Change from format "\<functionName\>: \(error)" to format "\<functionName\> failed with error: \(error)".

HirdayGupta commented on Bruin Bite/Network/UserAPI.swift:
```diff
@@ -119,10 +119,10 @@ extension API {
             switch result {
             case let .success(response):
-                print("Delete: \(response)")
+                Logger.shared.handle(type: .info, message: "User Deleted: \(response)")
```
Good precedent for the `.info` type! :)

HirdayGupta commented on Bruin Bite/Network/UserAPI.swift:
```diff
@@ -135,16 +135,16 @@ extension API {
                     let results = try JSONDecoder().decode(MatchRequestData.self, from: response.data)
                     let jsonData = JSON(response.data)
-                    print(jsonData)
+                    Logger.shared.handle(type: .debug, message: "matchUser JSON data: \(jsonData)")
```
Good precedent for the `.debug` type!

HirdayGupta commented on Bruin Bite/Network/UserAPI.swift:
```diff
@@ -135,16 +135,16 @@ extension API {
+                    Logger.shared.handle(type: .error, message: "matchUser: \(err)")
                 }
-                print("Successfully Sent match")
+                Logger.shared.handle(type: .verbose, message: "Successfully Sent match")
```
Change to `.info`

HirdayGupta commented on Bruin Bite/UserClass.swift:
```diff
@@ -73,7 +73,9 @@ public class User {
             net_error = (user_error_model?.error_description)!
         }
-        print(net_error)
+        if(net_error != "") {
```
Can remove the check for empty, since our logger function already checks for that.

kamalca replied:
If net error is empty in this case, Logger will still log "Warning: Net Error:" because message is not empty. Which means we will be getting warning messages every single time we use this update error function. I put the empty check in the Logger class because an empty error should never be logged, not to handle cases like this.

HirdayGupta replied:
Okay cool, that sounds good!

HirdayGupta commented on Bruin Bite/View Controllers/MenuVC.swift:
```diff
@@ -77,7 +77,7 @@ class MenuVC: UIViewController, UICollectionViewDelegate, UICollectionViewDataSo
             self.topBar.timeSelected("Night")
         default:
-            print("Current Meal Period could not be determined")
+            Logger.shared.handle(type: .error, message: "Current Meal Period could not be determined")
```
Can you reconfirm if this needs to be `.error` or if we can change it to `.warning` (read the aforementioned descriptions for the intention of each type, basically, if the meal period not being determined breaks things in a very noticeable fashion then leave it as `.error` but if not, and its a minor inconvenience, then change to `warning`)

kamalca replied:
I believe this wold result in the behavior we were seeing last year where the menu would not load until you clicked on a different time. However, this switch statement covers all possible values of hour from zero (inclusive) to 24 (exclusive). Therefore the error is more of a formality. I can switch it to warning though.

HirdayGupta commented on Bruin Bite/View Controllers/PopupViewControllers/TimePickerViewController.swift:
```diff
@@ -67,7 +67,7 @@ class TimePickerViewController: UIViewController {
         }
     } else {
-        print("No meal period selected")
+        Logger.shared.handle(type: .warning, message: "No meal period selected")
```
Confirm if this needs to be `.error` or can stay as a `.warning`

kamalca replied:
This warning/error occurs when someone tries to select a time before selecting a meal period. The resulting behavior is simply a popup that is not populated with times. I would not consider it a fatal error by any means. I suggest keeping it a warning and later alerting the user to this error when they attempt to put in a time before the meal period.

HirdayGupta reviewed (changes requested):
Hey! Sorry for the delayed response, great job adding in the new class and calling it instead of all of our (overly generous) uses of the print statement! I think I agree with your ErrorType classifications on most of the print statements you replaced.

Leaving some requested changes that are mainly aesthetic in nature. Great work otherwise!

HirdayGupta commented:
@kamalca When there are many requested changes, Github hides some of the requested changes by default, so be sure to click the `show hidden` button to get every requested change! Also feel free to comment on any of the review comments to start a discussion about the changes requested if you feel you didn't understand a requested change or do not agree.

kamalca committed:
Completed Requested Changes

Carried out the changes requested by @hirdaygupta --
• Static functions instead of a singleton
• Changed parameter names and function name from handle to log
• Changed file from ErrorLogger to LogManager
• Changed enumeration name from ErrorType to LogLevel
• Removed raw values from enumeration and added a function to determine the header
• Added documentation to enumeration
• Revised some log messages and log levels

HirdayGupta commented on Bruin Bite/Supporting Files/AppDelegate.swift:
```diff
@@ -86,17 +86,18 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
-        print("Device Token: \(token)")
-        print(token)
+        Logger.log("Device Token: \(token)", withLevel: .debug)
+        Logger.log(token, withLevel: .debug)
```
Should be safe to remove `Line 90` because its a repeat.

HirdayGupta commented on Bruin Bite/Managers/LogManager.swift:
```diff
@@ -0,0 +1,63 @@
+
+class Logger {
+    
+    private init() {    }
```
Can remove the private initializer because an initialized `Logger` class has no use if all its functions are static (so even if a dev was to initialize a Logger object, it would be useless).

HirdayGupta commented on Bruin Bite/Managers/LogManager.swift:
```diff
@@ -0,0 +1,63 @@
+class Logger {
+    
+    private init() {    }
+    
```
Also, since we're trying to document our code better, I'd suggest adding the following comment block above this function (it's a swift documenting comment that will show up on Xcode autocomplete whenever someone tries to use this function).

```swift
/**
Prints the given message to the log pre-fixed by the header associated with the given log level.

- Parameters:
    - message: A string with the message that needs to be logged.
    - level: The LogLevel with which the message needs to be logged
*/
```

HirdayGupta reviewed (changes requested):
Whoops, left this out of my previous review.

HirdayGupta commented:
@kamalca Any progress on this Pull Request?

kamalca committed:
Merge branch 'development' into kameron_error_handling_class

kamalca committed:
Completed requested changes pt.2

made requested changes

HirdayGupta reviewed (approved):
Looks good, thanks for putting in the effort @kamalca! 😄
